### PR TITLE
Improve Terminology? - nullable primitive type

### PR DESCRIPTION
**Pull request by bgribaudo**
**Background:** 
Currently, `Type.Is`'s documentation gives no indication that its second argument should only be set to certain type values. A [recent PR](https://github.com/MicrosoftDocs/query-docs/pull/200) attempted to rectify this by updating `Type.Is`'s docs, copying over and expounding on the rule [given in the spec](https://learn.microsoft.com/en-us/powerquery-m/m-spec-types#type-equivalence-and-compatibility) which states that `Type.Is` only accepts a "nullable primitive type value as its second argument." 

**Problem:**
However, when reviewing the PR, @ehrenMSFT pointed out that `Type.Is` works fine with second argument values such as `type number`, which clearly is not nullable (e.g. `Type.IsNullable(type number) // returns false`). 

So, the spec says that `Type.Is`'s second argument must be a _nullable_ primitive type, yet the function works fine with primitive type values that are not nullable.

I think the reason for this dichotomy is in the definition of terms:
* The spec [is using](https://learn.microsoft.com/en-us/powerquery-m/m-spec-consolidated-grammar#is-expression) "nullable primitive type" to refer to the set of "all primitive types" _plus_ their nullable counterparts (so would include `type number`).
* While a literal reading of the term "nullable primitive type" can be interpreted to imply a primitive type value that passes the `Type.IsNullable` test (so would exclude `type number`).

**Question:**
To avoid/resolve this confusion, should a different term be used in the spec for the set "all primitive types + their nullable counterparts"?

(Note: The term nullable primitive type is used a number of places in the spec, including in the details of the `as` and `is` operators as well as in conjunction with the definition of function parameters. So, this change would affect a number of places in the spec.)

### CODE REVIEW

**Comment by Court72** [comment]
@DougKlopfenstein

Can you review the proposed changes? 

When the changes are ready for publication, add a `#sign-off` comment to signal that the PR is ready for the review team to merge. 

#label:"aq-pr-triaged"
@MicrosoftDocs/public-repo-pr-review-team

**Comment by egorelik93** [comment]
[bgribaudo](https://github.com/bgribaudo) I suspect you are right about how the terminology was being used here, and I agree that this should be rephrased. I don't have a strong opinion on what we should change it to. The simplest might be to just say "primitive and nullable primitive types". I am also thinking of "optionally nullable primitive types", but this seems like it could still be confusing.

**Comment by bgribaudo** [comment]
Thanks, @egorelik93!

What about calling these "non-custom types"?

This would correspond nicely with (e.g. it would be the exact inverse of) what the [spec calls custom types](https://learn.microsoft.com/en-us/powerquery-m/m-spec-types): "All types that are not members of the closed set of primitive types plus their nullable counterparts are collectively referred to as custom types."

It seems like it would read clearly. Example:
> The second argument to `Type.Is` should be a non-custom type value.

**Comment by bgribaudo** [comment]
Another idea could be the name "simple types".

**Comment by bgribaudo** [comment]
@egorelik93, hope your day is going well! Did you have any thoughts on this?

**Comment by egorelik93** [comment]
@bgribaudo I take that back. 'Built-in types' could ostensibly include things like Int64.Type, which we don't want to go into. If 'primitive and nullable primitive' is too wordy, maybe let's just use 'non-custom' types.

**Comment by ehrenMSFT** [comment]
> primitive and nullable primitive

I think "primitive and nullable primitive" would be fine. It seems like that's the clearest option.

I'd also be fine with just saying "primitive", since most people would probably assume that includes the nullable versions of those types as well.

But if you want to be explicit and say "primitive and nullable primitive", I'm fine with that.

**Comment by ehrenMSFT** [comment]
I'm fine with either "primitive and nullable primitive" or "non-custom".

File: query-languages/m/m-spec-types.md [hunk 1]
```diff
@@ -209,28 +211,28 @@ A conforming value may contain field names not listed in the field specification
...
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`optional` _parameter-specification<br/> 
 parameter-specification:<br/>
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;parameter-name parameter-type<br/> 
-function-return-type:<br/>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;parameter-name  parameter-type<br/>
+parameter-type:<br/>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;assertion<br/>
 assertion:_<br/>
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`as`  _nullable-primitive-type_
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`as` _type_
```
Comment (bgribaudo) [inline query-languages/m/m-spec-types.md:235]: I believe this is correct, as variables, etc., can be used in an `as` clause in a function type definition.

**Comment by bgribaudo** [comment]
Thank you, @ehrenMSFT and @egorelik93.

Changes ready for review.

Notes: 
* These changes also incorporate [a correction[(https://github.com/MicrosoftDocs/query-docs/pull/203#discussion_r1638951983) to the grammar rule for function type definitions.
* Some of the function signature grammar rule names were out of sync between the consolidated grammar and the functions page. Resolved.

**Comment by bgribaudo** [comment]
Hi @egorelik93 and @ehrenMSFT! Hope you both had a great holiday! Wanted to circle back on this PR. Are we good to go here?

**Comment by bgribaudo** [comment]
Hi @egorelik93, @ehrenMSFT and @DougKlopfenstein!

Hope you each are doing well! 

Wanted to circle back on this PR. Seems like there was consensus with the idea of the naming change, but there hasn't been any review activity in a long time. Are we good to move forward with this PR? :-)

**Comment by DougKlopfenstein** [comment]
@bgribaudo, @ehrenMSFT, @egorelik93 - Is this pull request ready to be merged?

**Comment by egorelik93** [comment]
@bgribaudo I'm leaning back towards "primitive and nullable primitive". "non-custom" is fine for spec wizards but in isolation would be non-obvious.

**Comment by bgribaudo** [comment]
Wouldn't someone have to be a bit familiar with the spec to know what the term "primitive" means, as well?

**Comment by bgribaudo** [comment]
@ehrenMSFT, @egorelik93, what do you think of this? :-)

**Comment by bgribaudo** [comment]
Hi @egorelik93 and @DougKlopfenstein! Happy New Year! Do you know the status of this PR? Thanks!

File: query-languages/m/m-spec-operators.md [hunk 2]
```diff
@@ -1278,13 +1278,15 @@ The type assertion operator `x as y` is defined for the following types of value
 
 | X | Y | Result |
 | --- | --- | --- |
-| `type any` | _nullable-primitive-type_ | `type any` |
+| `type any` | _primitive-or-nullable-primitive-type_ | `type any` |
 
-The expression `x as y` asserts that the value `x` is compatible with `y` as per the `is` operator. If it is not compatible, an error is raised. `y` must be a _nullable-primitive-type_.
+The expression `x as y` asserts that the value `x` is compatible with `y` as per the `is` operator. If it is not compatible, an error is raised. `y` must be a _primitive-or-nullable-primitive-type_.
```
Comment (egorelik93) [inline query-languages/m/m-spec-operators.md:1283]: I don't want us creating 'primitive-or-nullable-primitive-type' as an official atomic piece of terminology. It will be in the grammar because we have to implement the grammar somehow, but we may still yet decide on a better name for this group later. In prose, we just want to refer to the union of two sets that already exist: the primitive types, and the (actual) nullable primitive types.

File: query-languages/m/m-spec-types.md [hunk 3]
```diff
@@ -26,24 +26,26 @@ A _type value_ is a value that _classifies_ other values. A value that is classi
...
 
-All types that are not members of the closed set of primitive types plus their nullable counterparts are collectively referred to as _custom types_. Custom types can be written using a `type-expression`:
+The set of _primitive types_ plus their nullable counterparts are known as the _primitive-or-nullable-primitive-types_ or alternately as the _non-custom types_.
```
Comment (egorelik93) [inline query-languages/m/m-spec-types.md:29]: Keeping with that theme, let's not change the prose in this part.

**Comment by bgribaudo** [comment]
@egorelik93, thanks for the feedback. Any better now?

File: query-languages/m/m-spec-operators.md [hunk 4]
```diff
@@ -1253,21 +1253,21 @@ The type compatibility operator `x is y`  is defined for the following types of
 
 | X | Y | Result |
 | --- | --- | --- |
-| `type any` | _nullable-primitive-type_ | `type logical` |
+| `type any` | _primitive-or-nullable-primitive-type_ | `type logical` |
 
-The expression `x is y` returns `true` if the ascribed type of `x` is compatible with `y`, and returns `false` if the ascribed type of `x` is incompatible with `y`. `y` must be a _nullable-primitivetype_.
+The expression `x is y` returns `true` if the ascribed type of `x` is compatible with `y`, and returns `false` if it is not compatible. `y` must be a _primitive-or-nullable-primitive-type_.
```
Comment (egorelik93) [inline query-languages/m/m-spec-operators.md:1258]: Still some prose references to 'primtive-or-nullable-primitive-type' around here.

File: query-languages/m/m-spec-types.md [hunk 5]
```diff
@@ -29,22 +29,24 @@ The set of _primitive types_ includes the types of primitive values, and a numbe
 
 All types that are not members of the closed set of primitive types plus their nullable counterparts are collectively referred to as _custom types_. Custom types can be written using a `type-expression`:
 
+_primitive-type:_ one of<br/>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`any anynonnull binary date datetime datetimezone duration function list logical`<br/>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`none null number record table text time type`<br/>
+_primitive-or-nullable-primitive-type:_<br/>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`nullable`_<sub>opt</sub>  primitive-type_<br />
+
+All other types are collectively referred to as _custom types_. Custom types can be written using a `type-expression`:
```
Comment (egorelik93) [inline query-languages/m/m-spec-types.md:38]: Is this supposed to be here? Seems redundant with the previous prose paragraph.

File: query-languages/m/m-spec-operators.md (hunk 4 above)
Comment (bgribaudo) [reply query-languages/m/m-spec-operators.md]: Fixed.

File: query-languages/m/m-spec-types.md (hunk 5 above)
Comment (bgribaudo) [reply query-languages/m/m-spec-types.md]: Fixed. Thanks!

File: query-languages/m/m-spec-functions.md [hunk 6]
```diff
@@ -26,24 +26,24 @@ parameter-list:<br/>
...
 parameter-type:<br/>
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;assertion<br/>
-function-return-type:<br/>
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;assertion<br/>
-assertion:_<br/>
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`as`  _nullable-primiitve-type<br/>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;primitive-or-nullable-primitive-type-assertion<br/>
+return-type:<br/>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;primitive-or-nullable-primitive-type-assertion<br/>
+primitive-or-nullable-primitive-type-assertion:_<br/>
```
Comment (egorelik93) [inline query-languages/m/m-spec-functions.md:38]: This is as convincing as anything that we really should have a shorter name for this concept. I'm not going to block the change on that though.

File: query-languages/m/m-spec-operators.md [hunk 7]
```diff
@@ -133,12 +133,12 @@ The following table summarizes the M operators, listing the operator categories
     <tr>
         <td>Type assertion</td>
         <td><em>x</em> <code>as</code> <em>y</em></td>
-        <td>Is compatible nullable-primitive type or error</td>
+        <td>Test if value is compatible with primitive type or nullable primitive type</td>
```
Comment (egorelik93) [inline query-languages/m/m-spec-operators.md:136]: In the original, this is not the same text as for "is"

File: query-languages/m/m-spec-types.md [hunk 8]
```diff
@@ -210,28 +209,28 @@ A conforming value may contain field names not listed in the field specification
...
 optional-parameter-specification:_<br/>
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`optional` _parameter-specification<br/>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`optional` _parameter-specification<br/> 
 parameter-specification:<br/>
-&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;parameter-name parameter-type<br/> 
-function-return-type:<br/>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;parameter-name  parameter-type<br/>
+parameter-type:<br/>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;assertion<br/>
 assertion:_<br/>
```
Comment (egorelik93) [inline query-languages/m/m-spec-types.md:232]: For consistency with your rename in functions from "primitive assertion" to "primitive or nullable primtive type assertion", maybe we should rename this as "type assertion". I'm also open to leaving them both as "primitive assertion" and "assertion", respectively.

Comment (bgribaudo) [reply query-languages/m/m-spec-types.md]: Change made. Thanks!

File: query-languages/m/m-spec-operators.md (hunk 7 above)
Comment (bgribaudo) [reply query-languages/m/m-spec-operators.md]: Is now better? (Much closer to original.)

A catch with this wording is that it could be read as checking compatibility with either a type _or_ an error.

**Comment by bgribaudo** [comment]
Hi @egorelik93! Did you need anything else from me on this?

File: query-languages/m/m-spec-operators.md [hunk 9]
```diff
@@ -1253,21 +1253,21 @@ The type compatibility operator `x is y`  is defined for the following types of
...
+primitive-or-nullable-primitive-type:_<br/>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`nullable`_<sub>opt</sub> primitive-type_<br />
 
 Type compatibility, as supported by the `is` operator, is a subset of [general type compatibility](m-spec-types.md) and is defined using the following rules:
 
-* If `x` is null then it is compatible if `y` is the type `any`, the type `null`, or a nullable type.
+* If `x` is null then it is compatible if `y` is the type `any`, the type `null`, or a nullable primitive type.
 
-* If `x` is non-null then if it is a compatible if the the primitive type of `x` is the same as `y`.
+* If `x` is non-null then it is a compatible if the primitive type of `x` is the same as `y`.
```
Comment (egorelik93) [inline query-languages/m/m-spec-operators.md:1270]: Can you remove the 'a' to fix the grammar?

**Comment by egorelik93** [comment]
There's a small nit, but otherwise I think this looks good. @DougKlopfenstein, do you have any additional thoughts?

**Comment by DougKlopfenstein** [comment]
@egorelik93 @bgribaudo - All the non-technical changes look good except for your minor nit. If that's fixed, everything else looks good.

File: query-languages/m/m-spec-operators.md (hunk 9 above)
Comment (bgribaudo) [reply query-languages/m/m-spec-operators.md]: Good catch. Thanks! 

Done.

**Comment by bgribaudo** [comment]
@DougKlopfenstein, minor grammar error fixed!

**Comment by DougKlopfenstein** [comment]
@egorelik93 - Please approve the changes and then I can merge the pull request.

**Review by egorelik93 — APPROVED** [review]

**Comment by bgribaudo** [comment]
@egorelik93 and @DougKlopfenstein, thank you for your help with this PR (and your patience with it!). Much appreciated.
